fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1315,7 +1315,7 @@ def _post_reduction_hook(self, param: Parameter, reduced_grad: torch.Tensor) ->
         # Optionally move gradients to CPU, typically used if one is running
         # the optimizer on the CPU.
         if self.move_grads_to_cpu:
-            param._cpu_grad.copy_(param.grad.data, non_blocking=False)
+            param._cpu_grad.copy_(param.grad.data, non_blocking=True)
```

Comment:
Ideally we would use non-blocking GPU -> CPU transfer here. This was originally enabled, but I had to disable it in #496 because I found it was producing non-identical results.

I was not able to build a unit test to reproduce this problem, but it happens with the fairseq example here: https://github.com/pytorch/fairseq/tree/master/examples/fully_sharded_data_parallel#13b-params-on-1-v100-gpu-with-cpu-offloading

Basically, you should see identical results with non_blocking=True and non_blocking=False, but I didn't see identical results at the time. This suggests a missing stream synchronization or something similar.

Probably this will be fixed by an extra `torch.cuda.synchronize()` in fairseq's cpu_adam.py, which is being added here: https://github.com/pytorch/fairseq/pull/3650/files#diff-a7721b9e9db5b90f71bb01c359c13835fd98d30562e5af9127cb388d53ea65c7

We should confirm that combining these two changes (non_blocking=True + torch.cuda.synchronize() at in cpu_adam.py) will yield identical results.

Reply:
Hi Myle,
I get identical loss for 8 GPUs 30 iterations,dummy_lm, with and without these changes, on gpt-13, full offloading and  checkpointing, cpu_adam (with that stream synchronization).

WO
```
I0628 140700.318 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3120.13 | wps 2669.2 | ups 0.02 | wpb 131072 | bsz 64 | num_updates 30 | lr 0.0001 | gnorm 3.806 | loss_scale 2 | train_wall 1726 | gb_free 16.7 | wall 1794
```           

With:
```
I0628 141512.471 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3120.13 | wps 2322.4 | ups 0.02 | wpb 131072 | bsz 64 | num_updates 30 | lr 0.0001 | gnorm 3.806 | loss_scale 2 | train_wall 1953 | gb_free 16.7 | wall 2022
```

As well as on a single GPU:

WO:
```
I0628 132805.584 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3119.93 | wps 125.1 | ups 0.01 | wpb 16384 | bsz 8 | num_updates 30 | lr 0.0001 | gnorm 3.387 | loss_scale 4 | train_wall 4286 | gb_free 16.7 | wall 4357
```

With:
```
I0628 133745.308 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3119.93 | wps 133 | ups 0.01 | wpb 16384 | bsz 8 | num_updates 30 | lr 0.0001 | gnorm 3.387 | loss_scale 4 | train_wall 4011 | gb_free 16.7 | wall 4084
```

Reply:
Thanks, looks awesome!

Reply:
Does the WPS (words per second) from the logs above suggest that this slows down training?

### My interpretation
- I assumed "WO"=> "Without these changes" => master
```
N-gpu/master: 2669.2
N-gpu/branch: 2322.4 

1-gpu/master: 125.1 
1-gpu/branch: 133
```
=> this change introduces a 14% slowdown (2669/2332-1) for the N-GPU case and a 6% speedup for the 1 GPU case.

Reply:
Question related to the torch.cuda.synchronize() fix. Is this something we need to do in the step() function of any optimizer? Should we test this with non-fairseq code?

Reply:
Nope, only if we are offloading to CPUs, because there is nothing that prevents CPU from getting ahead and not waiting for the GPUs to transfer the proper gradients.

Reply:
If the optimizer is on GPU, the stream semantics ensures the ordering (as long as the optimizer and the gradient are processed on the same stream)